Add new standarized difference diagnostics

ivanhigueram opened the pull request:
Did the 450B HW as suggested in #1 . A new method `diagnose` returns a "love" plot with the Standardized  Mean Differences or the Absolute Standardized  Mean Differences using the `metric` argument. I want to add something to compare the whole distribution, like a KS tests with weights, but I have to think more about that. You can access to the hidden method `self._covariate_differences`. 

Also, the function can now take a `pd.DataFrame` instead of the `np.ndarray`, just to keep things nice in the plot naming. If a `pd.DataFrame` is not passed, it will label things in a naive way, see the notebook for an example. 

I added some typing stuff, not that really matters.

ivanhigueram committed:
Add new standarized difference diagnostics

Keeping everything in pytorch and do a standarized differences
diagnostics plot for the CBPS class. Also, change a bit the input
of the class to take a dataframe too.

Update notebook with the new stuff.

Bugs:
 - I haven't checked things thoroughly
 - I want to implement a ks stat.

apoorvalal commented on src/diagnostics/differences.py:
```diff
@@ -0,0 +1,88 @@
+        mean_control = X[treat == 0, :].mean(axis=0)
+
+    # Calculate for all the variables
+    std_vars = X.std(axis=0)
```
instead of pooling the variance across treatment and control, the advice in Imbens and Rubin is to compute $\sqrt{(\sigma^2_1 + \sigma^2_0)/2}$ where $\sigma^2_w$ is the variance of $X$ in treatment group $w$. This is particularly relevant in super asymmetrical T/C groups and lots of heteroskedasticity.

apoorvalal replied:
did a bunch of other stuff but forgot to do this, haha. Lmk if you get around to it, otherwise i'll do it before merging.

apoorvalal committed:
added third option to diagnose to plot a scatterplot of standardized vs unstandardized balance stats, and added an example with high dimensional data. Also added data dependencies.

apoorvalal commented on src/cbps_torch.py:
```diff
@@ -139,38 +139,48 @@ def _covariate_differences(self, metric):
             matplotlib.axes
         """
-        fig, ax = plt.subplots(figsize=(10, 6))
+        if not ax:
```
added option to pass `ax` so that figures can be combined

ivanhigueram replied:
oh this is way nicer!

apoorvalal commented:
Thanks for following through on that belated problem set hw :smile: ; nice work. I added a third option for high-dim data that just scatters unweighted-on-weighted mean differences (where you want everything to be below the 45 degree line, and hopefully as close to 0 as possible). Feel free to refactor; I think you have a good eye for this sort of thing.

ivanhigueram committed:
Change weights to return numpy array if wanted

Small changes to CBPS class to return numpy array if desired. Also, add
a new way to calculate variance when doing diagnostics.

ivanhigueram commented:
Think the plot is ok, maybe normalizing the weights to they're bounded (0, 1), but not sure if that would be informative. 

Made some small changes: 
 - Solves #3 
 - Calculate the standard deviation correctly using your advice!

apoorvalal reviewed (approved):
Looks great. This is starting to look like a real library now.
